### DISCUSSION TITLE
We need your help to implement the `pipe` function 🙏

### DISCUSSION THREAD

**Discussion by fabian-hiller** [open-circle/valibot#498] [Announcements]
Based on #463 I started to rewrite Valibot and implement the `pipe` function. The results so far are promising. I was able to reduce the bundle size, improve type safety and fix some general issues. Unfortunately, I hit a [TypeScript limitation or bug](https://github.com/microsoft/TypeScript/issues/57978). My current implementation is completely typesafe. Even the issues are now fully typed. So I want to avoid compromising here at all costs.

Is there a TypeScript wizard in our community like @mattpocock or @ssalbdivad that can find a workaround? I have created a [TypeScript playground](https://www.typescriptlang.org/play?#code/PTAEBcE8AcFMGcB056gLQc17Pd-wYWgFDECWAduLAE4BmAhgMaygBCD8sAykwBawAtgwA8AFQCSFaAFdwAPlABvYqFABrSgBMAXKADk8fkIb6A3KogxYe+OBqUA5hbUhQAWQD2NVtBqfoVAEfRFDLAH0aGQoACkpZcD1o9QpPAHcKAEo9SWk5F1BwqDh4AH49JVB4uRypBLNQTzkE2rzwUABfCw7SSmp6ZlYOLgBBJnAyTwpxOrkAGlAxAHlmuUUVNU0KXQMANwYAGzItBgmp-VAAHwN7Bgp4Om9hM4pzS2KbUDsHCmdLNy8PlAfgCQVosFCiAiUVi1USi1m4GyixW4HqEQ+ZQqVTarXqjVW8OWhM63VIH1AUjotER4l4AmEoFgAA9qNtUMMeMZhCI7pBFNdOWMXryKJAFnz5IoALyWAByUzlMgOBwYACMDrA6dyGABtfRFazwfQAXXk+rhposxApVNoqIS2oZDCZrNg7PYnC5ztF-KuntG40m0z5ErFUtAsrUCooSpV6s1TpM+sNJVN5v0TTRcitpDc3ycyFQRBLpbLRF6VFojBYoG49ic9JMrrZWg5XqbPILv3W72sekMDd+bx6xDo0SDUy+Q8cMWR9Z+jk7Lo2oB84BkNAoyksm20A6Mzv0c13VjgB5nx9PkWicTamR3aifoDcYj4ZFQH9AACsZHYvgwgjQJqoBMJ4WiwKeahkHQoAxB8niwXCoAAITStKBjdo4+gPquz5qOAfD+GkoAULAJEAKI0P4NAxPoUj7EcWhnrAOEFM+PT4WusAbluOLohxJ5qF0xCjm4giUAAMu6jiEUW6DlopSkVuQVYDLW7hSTJhHCsGIiWLkCQtu6bbTou-p8rqJpCYsABKsAAI4yGQPiCO67Qsq2qAUDIghqrQJ6KJ5JntoGIqGfMCJtL2e7bAOjHHKcwZvAR-YGBJFDhJqvyESl3FOS5QjuTk9kFa57lkmOE4vKAGXSTlfAzG0xkelhFlilZCxiKVznlVQLWmT5fm0PIMSWD4ZVFVQJWOb1U3gMQyKaRQ9WyXwulTE1CRdT1hVuVQMXcbx254Ro+57IciUvFez4fAOGVZdpfA3U+E1zft4A2WoN6wvej74a+76fqgv7-vAgHAawYEQVBVSwXeCSINla2gCI+Xve5uGwwRRHpKR5GgFRNF0Qxl3McjuWZOxT6cfh66btucLUx0NkiWJYC3Pcjw0II8nKfzAtYJW-Q1qwYg0HcDxPBt0wRZ9KKEkFboekKk6y4iXUOms-1bDs+ic1LPNJecBR3TcEtc08I6kOOFBqxAFuG4IW2RcS2YKGNajMC8egIzUUUJA+0qKG7CSLTkjvc4IMsu-Lofa6u9N8adusDgbUfG68X0sWnkdW9nP1+0i-3PoDX5fqD7Tg0BIHQ5BXFJ9u3vBkXVOnizlhs3mYDQGQcB84Lg+C+S1h1jqADqZCEQACn3WoGcuA2hd6Ji+vINliLPcASNQgioMFHq6qrIqhqAkoLJCx96afkpWdZBla59xAyqASwSeASbCAsBowhc1wGpifQigABk-1e7nkWFvWAO8hDwAKIXOESQKApHSFkHIj94GYnKMofi-s7Q0FpGIZc8gGhZhaArd2pJRLWjcAqagegd5VG8rAFg8BwY0EgBAPgpxQAAANN5zxgYIAAjLwphZ9QCOHdLQMgTBECv0IrQNIH5WCKKOgzYgbgKSIT4eA2AYi0hNAOMxVI7R-KgHHCqThFJwbUihLbe2ej9JqCITqJeAYV48nPgZKBQjhHuKvptfBj9P4MHXmfcMgVPZfB1DkZcNk9F+JyL43ewjw5j2dJPGec9QkLF1AI7eqTrKUgoNSGgISCnQNSRGROPEGYl1AJCQ8Jhs56L0LqRJRSC4wlbg01wYA3zlxBn+KuENa7gXrnTOpfFOlCGEYgQuzThALJ6XCTIbdBKd0qm4YCDBICOH8NELQA8h6nMIN3UAABxTwhwcjWF4A4aAVc+BGJMZ4doYEa4MEoGfNUWYuGsEeCqdIThQITPkVIOwsAGBaAWFPTRYB4AvOVMxBgchPDPFkYcA4nDKBlIBSxRosE1GUDIjQCx1VgygDVF6ZiU41FZloAi3Rc8LgOJeFCYgYF7jtGXP4jCeiYhYTnAsOqT0YgAFZ1nWm5f+ZcAAmSMwI57ROFZkGy6cngxF6cHXB4AkZPUyItCwQA) with a minimal reproduction.

**Comment by ssalbdivad** [comment]
I took a quick look, and if I'm understanding the API correctly, the first issue can be resolved by ensuring the input type is correctly treated as contravariant so a broader type like `string | any[]` can be assigned to a narrower one like `string`. I made a couple adjustments to demonstrate that here:

https://tsplay.dev/WJ3lgw

However, you will run into a limitation inferring piped output with this approach. Without using HKTs, you will not be able to have a function like `minLength` that will return either a string or an array depending on the type of the input. You would have to separate it into two separate constraints, but you'd essentially run up against the same issue if you ever needed a narrowed instantiation of one of those types like `number[]`.

The options for working around this are to define your actions as HKTs like those in HotScript or hkt-toolbelt so they can be applied at a type-level and return a result that depends on the input, or to use a different API. ArkType's next release uses chaining to handle this scenario:

![image](https://github.com/fabian-hiller/valibot/assets/10645823/654d7bb8-13f3-4fb4-9cf2-cb36d7bcd752)

Hopefully something like that could be made compatible with your tree-shaking strategy?

**Reply by fabian-hiller** [reply]
Thank you David for taking the time! Unfortunately, your solution does not solve the root of our problem. It only removes the TS error, but at the same time it changes the output type of `Schema1` to `string | any[]` and that causes other problems.

**Reply by ssalbdivad** [reply]
Yes sorry my goal was not to solve the entire problem, but just to highlight what I understand to be an eventual limitation you'll encounter with this API.

Assuming you want to pipe some arbitrary list of transformations to some input and infer the narrowed output at each step, this cannot be achieved without some kind of HKT-based strategy.

I apologize if I misinterpreted the intent of the API.

I know you spoke with @Andarist about inferring inputs from context within the transform function. This is not a strategy I've ever used. To me, inference feels much more intuitive if the scopes are always handled bottom-up.

Have you considered removing the wrapper around the transform function altogether and instead interpreting a function passed directly to `pipe` as a transform? You could essentially "parse" out the same result.

**Reply by fabian-hiller** [reply]
Thank you for your feedback. I will think about it!

**Comment by brenelz** [comment]
Does this work the way you intend? Seems like the types flow pretty good but could just be accidental

https://tsplay.dev/NBdlxW

**Reply by fabian-hiller** [reply]
Thank you for your idea! The problem with this solution is that we lose type safety. `Schema1` now contains a `.pipe` with a `BaseAction` instead of a `MinLengthAction`. For example, this prevents us from accessing `Schema1.pipe[0].requirement` in a type-safe way.

**Reply by brenelz** [reply]
ah i see. I figured I was probably missing something

**Comment by anuraghazra** [comment]
Does this work? 

[TSPlay](https://www.typescriptlang.org/play?#code/PTAEBcE8AcFMGcB056gLQc17Pd-wYWgFDECWAduLAE4BmAhgMaygBCD8sAykwBawAtgwA8AFQCSFaAFdwAPlABvYqFABrSgBMAXKABE8fkIb6A3KogxYe+OBqUA5hbUhQAWQD2NVtBqfoVAEfRFDLAH0aGQoACkpZcD1o9QpPAHcKAEo9SWk5F1BwqDh4AH49JVB4uRypBLNQTzkE2rzwUABfCw7SSmp6ZlYOLgBBJnAyTwpxOrkAGlAxAHlmuUUVNU0KXQMANwYAGzItBgmp-VAAHwN7Bgp4Om9hM4pzS2KbUDsHCmdLNy8PlAfgCQVosFCiAiUVi1USi1m4GyixW4HqEQ+ZQqVTarXqjVW8OWhM63VIH1AUjotER4l4AmEoFgAA9qNtUMMeMZhCI7pBFNdOWMXryKJAFnz5IoALyWAByUzlMgOBwYACMDrA6dyGABtfRFazwfQAXXk+rhposxApVNoqIS2oZDCZrNg7PYnC5ztF-KuntG40m0z5ErFUtAsrUCooSpV6s1TpM+sNJVN5v0TTRcitpDc3ycyFQRBLpbLRF6VFojBYoG49ic9JMrrZWg5XqbPILv3W72sekMDd+bx6xDo0SDUy+Q8cMWR9Z+jk7Lo2oB84BkNAoyksm20A6Mzv0c13VjgB5nx9PkWicTamR3aifoDcYj4ZFQH9AACsZHYvgwgjQJqoBMJ4WiwKeahkHQoAxB8niwXCoAAITStKBjdo4+gPquz5qOAfD+GkoAULAJEAKI0P4NAxPoUj7EcWhnrAOEFM+PT4WusAbluOLohxJ5qF0xCjm4giUAAMu6jiEUW6DlopSkVuQVYDLW7hSTJhHCsGIiWLkCQtu6bbTou-p8rqJpCYsABKsAAI4yGQPiCO67Qsq2qAUDIghqrQJ6KJ5JntoGIqGfMCJtL2e7bAOjHHKcwZvAR-YGBJFDhJqvyESl3FOS5QjuTk9kFa57lkmOE4vKAGXSTlfAzG0xkelhFlilZCxiKVznlVQLWmT5fm0PIMSWD4ZVFVQJWOb1U3gCeyKaRQ9WyXwulTE1CRdT1hVuVQMXcbx254Ro+57IciUvFez4fAOGVZdpfA3U+E1zftC3XjCd4JLhUEvmAb5fl+v7-vAgHAawYEQf9MFwXCiDZWtoAiPl73uX9XEEUR6SkeRoBUTRdEMZdzFI7lmTsU+nH4eum7bnCVMdDZIliWAtz3I8NCCPJyl8-zWCVv0NasGINB3A8TwbdMEULSihJBW6HpCpOMuIl1DprI+Z1xTc4uc08SXnAUd16xLXOCCOpDjhQqsQPrkvc1tkXEtmChjWozAvHoP01FFv2RoorvbcQyJi+bUuq87cvB1rq503xp1bDs+gc47zzJTZqXnmbBvcy9ag3rC97a8+r7vp+qCg+04NASB0OQVxCfbl7wa+0iTMs5VbjQGQcC8wLg8C+S1h1jqADqZCEQACn3WoGcuA2hd6Ji+vINliLPcASNQgioMFHq6irIqhqAkoLJCx96afkpWdZBma59MqgEsEngEmwgLAaMIXNcBqYvoRQAAybWvcc6bznjvIQ8AChF3bkkCgKR0hZByI-WBmJyjKH4n7O0NBaRiGXPIBoWYWjyzdqSUS1o3AKmoHoHeVRvKwBYPAcGNBIAQD4KcUAAADCB29d4AEZuEMLPqARw7paBkCYIgF+hFaBpA-KwORR16bEDcBSRCPCwGwGEWkJoBxmKpHaP5UA44VTsIpODakUIbZ220fpNQBCdRLwDCvHk58DJb1gFAwQAiXFX02rg-BhCwz8kCh7L4OocjLhstonxAichePif6AJ0wgltA-gwdelIKDUhoI-TJUpFp6GXJPGec9MkLF1Hw7xgjrI5LyQUmp8SIzxx4vTUuoBISHhMFnYEc89C6jiXUvpcC4SY3wuXYGVc-w1whvXcCjdabtL4sMoQAjEBwJ6cITZ31xmU1PMzSwHQz6oFKVPPgXjKmgGqUkkZDT7SEnEHc9ZUpu5gGAgwSAjh-DRC0APIegLCB5jAAAcU8IcHI1heAOGgDXPg+jDGeHaGBOuDBKBnzVFmDhrBHgqnSE4UCiyZFSDsLABgWgFhTzUWAeACLlTMQYHITwGcmCHAOOwygeScUsUaLBZRlAyI0FMdVYMoA1RemYlOZRWZaA0q0XPAA5CK22LwoTEDAvcdoy4-EYW0TELCc4Fh1SejEAArJkA5Gqpj-gmpGMezoNnaN1AABhNIgN6e0KqkE1f+ZcAAme1+rLCGsyDZNOFsYjtwfNKRQCNyZ8DDaHCwQA)

- Schema1 returns `MinLength`, Can safely access `Schema1.pipe[0].requirement`
- No TypeErrors

<img width="768" alt="image" src="https://github.com/fabian-hiller/valibot/assets/35374649/fde8dad3-3ae6-4f19-871e-360f846c7eb6">


Let me know if I'm missing any edge case? 🤔

**Reply by Demivan** [reply]
Wow. Why does this work?

**Reply by fabian-hiller** [reply]
Like Ivan, I am blown away (and confused). 🤯😅
I will have a close look in the next few minutes to see if this workaround works for us.

**Reply by ssalbdivad** [reply]
@anuraghazra is a total boss 😎

**Reply by fabian-hiller** [reply]
The code is a little dirty, but it seems to work! 🥹 I am so grateful! 🙏 You probably saved all my work from the last few weeks. I was really afraid I would have to throw it all away. I will give another update this evening or at the weekend.

**Reply by anuraghazra** [reply]
I'm glad I was able to help. 😺 

> Wow. Why does this work?

To be honest even I'm not entirely sure. 

<img src="https://github.com/fabian-hiller/valibot/assets/35374649/5cbd3650-eb03-45fc-b9ee-4a75836affac" width="500px" alt="My code works and I don't know how meme" />

My guess is that by specifying the union of `BaseAction<InferOutput<TSchema>, InferOutput<TSchema>>` type in the argument, we are making the `pipeItem1`'s generic inference candidate much "better" thus TS is able to narrow down & return the concrete type. 

Here's a good resource on TS inference algorithm: https://portal.gitnation.org/contents/lets-make-a-generic-inference-algorithm

But of course my understanding is limited in terms of how the TS compiler is internally handling this case, maybe @Andarist can explain better.

**Reply by Andarist** [reply]
TS is able to infer from into return type of `minLength` based on the contextual type of this position. That contextual type is `TPipeItem1` if you don't use this union trick. So TS tries to instantiate this using the inferences that it was able to make so far. Nothing was inferred so far for `TPipeItem1` so it's instantiated with a magical "silent never type" and this type is internal and non-inferrable. The latter doesn't quite matter all of that much because it doesn't hold any useful information for us anyway but I guess it's non-inferrable to avoid spoiling inference results with it.

So nothing interesting happens at this stage and, I guess that, later on TS just pulls constraints and stuff when it touches this `minLength`.

But when you are using this union trick the contextual type is `TPipeItem1 | BaseAction<InferInput<TSchema>, InferOutput<TSchema>>`. It's also instantiated with the inferences made so far and while `TPipeItem1` is still instantiated with this magical silent never type it just gets eliminated from the union (`T | never` is just `T`). The other member of this union has better luck - it directly references type parameters that have some inference candidates already (`TSchema`). So this gets instantiated as `BaseAction<string, string>`. 

Based on that TS infers through the return type of `minLength` (from `BaseAction<string, string>` to `MinLengthAction<TInput, TRequirement>`). I didn't follow this further but I think this is essentially what "unblocks" it here. It's sort of a timing issue and an implementation detail leak. It's certainly an interesting effect but I'd call it somewhat accidental. I don't think there is anything fundamental stopping it from working better in this situation without relying on this trick.

**Comment by gmaxlev** [comment]
I've heard that it's not possible to implement the pipe function in TypeScript "normally" like in other languages. That's why the pipe function in RxJS was written like that [https://github.com/ReactiveX/rxjs/blob/master/packages/rxjs/src/internal/util/pipe.ts](https://github.com/ReactiveX/rxjs/blob/master/packages/rxjs/src/internal/util/pipe.ts)

**Reply by fabian-hiller** [reply]
I will take a look at it! Thank you!

**Reply by let-josh** [reply]
fp-ts is similar with their pipe function

https://github.com/gcanti/fp-ts/blob/a94d4f84d346da271e0450d7341b98e0a5739a0e/src/function.ts#L416

**Comment by fabian-hiller** [comment]
Feel free to take a look at the draft PR for the `pipe` function I just created: #502
